gammapy/image/models/core.py

```diff
@@ -50,9 +50,48 @@ def position(self):
             lon = self.lon_0.quantity
             lat = self.lat_0.quantity
             return SkyCoord(lon, lat, frame=self.frame)
-        except IndexError:
+        except AttributeError:
             raise ValueError("Model does not have a defined center position")
 
+    @property
+    def evaluation_radius(self):
+        """Evaluation radius of the model (`~astropy.coordinates.Angle`).
+
+        Can be a function of the model parameters, a constant value or None.
+
+        Examples
+        --------
+
+        from gammapy.image.models import SkyGaussian
+
+        gauss = SkyGaussian("0 deg", "0 deg", "0.2 deg")
+
+        # set evaluation radius as a function of model parameters
+        def eval_radius_gauss(sigma, **kwargs):
+            return 8 * sigma
+
+        gauss.evaluation_radius = eval_radius_gauss
```

Comment:
Why did you change this to `staticmethod _evaluation_radius` and dispatch from `evaluation_radius`? Why not keep as-is. Seems simpler and all use cases you have here can still be done.

Is it a good idea to teach people to monkey-patch `evaluation_radius` here in the docstring?
I mean, yes, experts can use monkey-patching throughout Gammapy to customise the analysis, but genereally it's not something we advertise in the docs, and I don't see why this evaluation radius should be something they need to monkey-patch.

Having a docs page that explains how to make custom models, usually from scratch or by sub-classing, very rarely by monkey-patching the existing ones, could be a nice addition.

Reply:
Yes, maybe we can find a better solution for this. My reasoning was as follows:

- The model evaluation currently uses bounding boxes and this we definitely want to keep 
- The `SkyDiffuseConstant` is not spatially confined, so I needed a way to switch off the bounding box evaluation for individual model components. I decided to allow for `evaluation_radius=None` to control, whether the bounding boxes are used or not. At this point we could stop and just implement fixed default values for all models (that's how it was, `evaluation_radius` was a property and not settable)
- But for some optimisers (e.g. sherpa's MC or similar) and models (e.g. `SkyDiffuseConstant`), it is useful to switch of the bounding box evaluation and I decided it would be a useful addition for to be able to define / switch-off the evaluation radius per model component
- I implemented a setter for `evaluation_radius`. But this is where the problem starts, because typically the evaluation radius is a function of some model parameters. And that's where the monkey patching is not avoidable, except if you pass a function that determines the evaluation radius on model init

I guess there are a few options:
- Make the evaluation_radius handling not configurable by the user
- Only allow for fixed evaluation radii, that do not depend on model parameters
- Invent a mini-language, so that users can pass `evaluation_radius="5 * {sigma}"`?
- Add an `evaluation_radius` argument for model init?
- Anything else...?

Reply:
I think having `None` as declarative value for the case of no evaluation radius is useful.

With the callable, you could just have kept everything the same, and very advanced users could have monkey-patched just by attaching a function with `self` as first argument. That's the common way to monkey-patch methods, no?

Reply:
Yes, I was struggling with the fact, that `evaluation_radius` was defined as a property and properties of instances can't be easily monkey-patched, I guess. But it's probably OK for power-users to do it on a class and not instance level?

Reply:
Like so:

```python3
from gammapy.image.models import SkyGaussian

def eval_radius(self):
    return 8 * self.sigma.quantity


SkyGaussian.evaluation_radius = eval_radius

# or to switch off
SkyGaussian.evaluattion_radius = None

```

Reply:
I don't recall if I ever monkey-patched a property at the instance level.

We have 1000s of properties in Gammapy ... is there a real use case to monkey-patch this one?

Reply:
To be honest, I don't want to spend much more time on this. I'll go back to the previous solution to have a global `evaluation_mode` argument on the `MapDataset` and remove the configuration per model component. The class-level monkey patching I've shown above is still possible...